File: app/routers/observations.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.orm import Session

from app.auth import require_api_key
from app.config import ANALYST_NAME
from app.db import get_db
from app.models import Company, MetricDefinition, Observation
from app.schemas.observation import ObservationBulkIn
from app.services.rule_engine import evaluate_observations

router = APIRouter(prefix="/api", tags=["observations"], dependencies=[Depends(require_api_key)])


@router.post("/companies/{company_id}/observations", status_code=status.HTTP_201_CREATED)
def post_observations(company_id: str, payload: ObservationBulkIn, db: Session = Depends(get_db)):
    company = db.get(Company, company_id)
    if company is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"company '{company_id}' not found")

    keys = {obs.metric_key for obs in payload.observations}
    known_keys = set(
        db.scalars(select(MetricDefinition.metric_key).where(MetricDefinition.metric_key.in_(keys))).all()
    )
    unknown = keys - known_keys
    if unknown:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"unknown metric_key(s): {sorted(unknown)}",
        )

    written = []
    for obs in payload.observations:
        stmt = (
            pg_insert(Observation)
            .values(
                company_id=company_id,
                period=payload.period,
                period_end=payload.period_end,
                metric_key=obs.metric_key,
                numeric_value=obs.numeric_value,
                text_value=obs.text_value,
                source_type=obs.source_type,
                source_url=obs.source_url,
                note=obs.note,
                ingested_by=ANALYST_NAME,
            )
            .on_conflict_do_update(
                index_elements=["company_id", "period", "metric_key"],
                set_={
                    "numeric_value": obs.numeric_value,
                    "text_value": obs.text_value,
                    "source_type": obs.source_type,
                    "source_url": obs.source_url,
                    "note": obs.note,
                    "ingested_by": ANALYST_NAME,
                },
            )
            .returning(Observation.id)
        )
        result = db.execute(stmt)
        written.append(result.scalar_one())

    db.commit()

    proposals = evaluate_observations(db, company_id, payload.period)

    return {
        "period": payload.period,
        "observation_ids": written,
        "count": len(written),
        "proposals": [
            {
                "id": p.id,
                "proposed_status": p.proposed_status,
                "source": p.source,
                "rationale": p.rationale,
                "evidence": p.evidence,
            }
            for p in proposals
        ],
    }

File: web/src/lib/server/services/observations.ts
// Ports app/routers/observations.py's service portion.
import { eq, inArray } from 'drizzle-orm';
import { db } from '../db';
import { companies, metricDefinitions, observations } from '../db/schema';
import { NotFoundError } from './scenarios';
import { evaluateObservations } from './ruleEngine';

export class UnknownMetricError extends Error {}

export interface ObservationIn {
	metric_key: string;
	numeric_value?: number | null;
	text_value?: string | null;
	source_type?: string | null;
	source_url?: string | null;
	note?: string | null;
}

export async function postObservations(
	companyId: string,
	period: string,
	periodEnd: string,
	items: ObservationIn[],
	actorIdentity: string
) {
	const [company] = await db.select().from(companies).where(eq(companies.companyId, companyId)).limit(1);
	if (!company) throw new NotFoundError(`company '${companyId}' not found`);

	const keys = [...new Set(items.map((o) => o.metric_key))];
	const knownRows = await db
		.select({ metricKey: metricDefinitions.metricKey })
		.from(metricDefinitions)
		.where(inArray(metricDefinitions.metricKey, keys));
	const known = new Set(knownRows.map((r) => r.metricKey));
	const unknown = keys.filter((k) => !known.has(k));
	if (unknown.length) {
		throw new UnknownMetricError(`unknown metric_key(s): ${JSON.stringify(unknown.sort())}`);
	}

	const written: number[] = [];
	for (const obs of items) {
		const [row] = await db
			.insert(observations)
			.values({
				companyId,
				period,
				periodEnd,
				metricKey: obs.metric_key,
				numericValue: obs.numeric_value != null ? String(obs.numeric_value) : null,
				textValue: obs.text_value ?? null,
				sourceType: obs.source_type ?? null,
				sourceUrl: obs.source_url ?? null,
				note: obs.note ?? null,
				ingestedBy: actorIdentity
			})
			.onConflictDoUpdate({
				target: [observations.companyId, observations.period, observations.metricKey],
				set: {
					numericValue: obs.numeric_value != null ? String(obs.numeric_value) : null,
					textValue: obs.text_value ?? null,
					sourceType: obs.source_type ?? null,
					sourceUrl: obs.source_url ?? null,
					note: obs.note ?? null,
					ingestedBy: actorIdentity
				}
			})
			.returning({ id: observations.id });
		written.push(row.id);
	}

	const proposals = await evaluateObservations(companyId, period);

	return {
		period,
		observation_ids: written,
		count: written.length,
		proposals: proposals.map((p) => ({
			id: p.id,
			proposed_status: p.proposedStatus,
			source: p.source,
			rationale: p.rationale,
			evidence: p.evidence
		}))
	};
}
